designs/node-upgrades.md

```diff
@@ -0,0 +1,185 @@
+
+#### Option 2b - In parallel with hard-coded hyper parameters and surface controls later
+
+While these defaults may make sense for some use-cases, it ultimately is tough to get right for others. This section could be a first step before implementing Node Disruption Budgets.
+
+Instead of introducing an API, we can upgrade nodes in parallel with a series of hard-coded defaults. Many of the similar defaults are implemented with [Cluster Autoscaler](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-are-the-parameters-to-ca)and allow configuration. Karpenter also uses [defaulted tunables in Consolidation](./consolidation.md#internal-tunables). For upgrades, we propose similar defaults:
+
+* Polling Period - This is how often a node will scan the cluster for nodes that need upgrading. Since this is fairly inexpensive to check with aws ec2 describe-instances, we can do this every 60 seconds. Otherwise, we can simply add the AMI-ID at provisioning time, where this becomes unnecessary.
+* Cooldown Period - How long Karpenter waits to upgrade instances after upgrading a batch. This may not be necessary as Karpenter will pre-spin nodes.
+    * This is similar to `scale-down-delay-after-add`, `scale-down-delay-after-delete`, and `scale-down-unneeded-time` in Cluster Autoscaler.
+* Minimum Node Lifetime - The minimum age a node must be before it is considered for upgrading. If a node is created and then a new AMI is released, it may be favorable to wait a bit before upgrading it as to not introduce a lot of churn. This can be mitigated with the following section, where we discuss upgrade ordering.
+* Max Upgrade Batch - How many nodes will be upgraded at a time. This will default to 5%. For a 1000 node cluster, 50 nodes could upgrade at the same time. For a cluster with 20 or less nodes, this will be 1, equivalent to upgrading serially.
+
+### Recommendation
+
+*Upgrade serially*. If we upgrade in parallel and use something like an NDB, users will need to use NDBs to safely onboard to automatic updates. Newer users may find this too complicated and disable automatic upgrades altogether. If Karpenter hard-codes defaults into the process, some cases may end up working poorly for existing users.
+
+Upgrading serially requires no additional API or hard-coded hyper-parameters. It creates simple requirements on upgrades, and solves some of the existing problems of jitter (https://github.com/aws/karpenter/issues/1884), more granular expiration conditions (https://github.com/aws/karpenter/issues/903), and rate-limiting (https://github.com/aws/karpenter/issues/1018). Upgrading serially is additionally extendable to various forms of the parallel upgrades as called out in each of the sections. Should users want quicker upgrades in the future, Karpenter can easily implement more controls where defaults retain the same assumptions.
+
+## What can interfere with upgrades?
```

Comment:
not specific to ICE but resource exhaustion in general - specifically with parallel upgrades (going back to worst case scenario of new AMI rolling out on large cluster) due to the theoretical doubling of resources during the reconciliation process:
- IP exhaustion
- AWS service limits (EC2 limits)

Reply:
Good callout. In a way, this makes a bigger case for upgrading serially, as we will only ever pre-spin one more node than the cluster's application's need (theoretically).